prov/sm2/src/sm2_rma.c

```diff
@@ -0,0 +1,509 @@
+		ret = sm2_write_err_comp(ep->util_ep.tx_cq,
+					 (void *) xfer_entry->hdr.context,
+					 xfer_entry->hdr.op_flags, 0,
+					 xfer_entry->hdr.cq_data, err);
+		if (!ret)
+			ctx->status_flags |= FI_SM2_SAR_STATUS_COMPLETED;
+		else
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Error in sm2_write_err_comp\n");
+	}
+	sm2_freestack_push(ep, xfer_entry);
+	if (ctx->msgs_in_flight == 0)
+		sm2_free_rma_ctx(ctx);
+}
+
+void sm2_fill_rma_ctx(struct sm2_ep *ep, const struct fi_msg_rma *msg,
+		      uint32_t op, uint64_t op_flags, sm2_gid_t peer_gid,
+		      struct sm2_sar_ctx *ctx)
+{
+	ctx->ep = ep;
+	ctx->msg.context = msg->context;
+	ctx->msg.rma_iov_count = msg->rma_iov_count;
+	ctx->msg.iov_count = msg->iov_count;
+	ctx->msg.addr = msg->addr;
+	ctx->msg.data = msg->data;
+
+	assert(msg->iov_count <= SM2_IOV_LIMIT);
+
+	memset(ctx->msg.desc, 0, msg->iov_count * sizeof(*msg->desc));
+	if (msg->desc) {
+		memcpy(ctx->msg.desc, &msg->desc,
+		       msg->iov_count * sizeof(*msg->desc));
+	}
+	memcpy(ctx->msg.msg_iov, msg->msg_iov,
+	       msg->iov_count * sizeof(*msg->msg_iov));
+	memcpy(ctx->msg.rma_iov, msg->rma_iov,
+	       msg->rma_iov_count * sizeof(*msg->rma_iov));
+
+	ctx->bytes_sent = 0;
+	ctx->bytes_acked = 0;
+	ctx->bytes_total = ofi_total_iov_len(msg->msg_iov, msg->iov_count);
+	ctx->msgs_in_flight = 0;
+	ctx->peer_gid = peer_gid;
+	ctx->op = op;
+	ctx->op_flags = op_flags;
+	ctx->status_flags = 0;
+}
+
+ssize_t sm2_rma_cmd_fill_sar_xfer(struct sm2_xfer_entry *xfer_entry,
+				  struct sm2_sar_ctx *ctx)
+{
+	ssize_t bytes_to_send, payload_used;
+	int ret;
+	void *src, *dest;
+	enum fi_hmem_iface iface;
+	uint64_t device;
+	struct sm2_rma_msg *msg;
+	struct sm2_cmd_sar_rma_msg *cmd_rma;
+	int jv;
+
+	msg = &ctx->msg;
+
+	xfer_entry->hdr.proto = sm2_proto_sar;
+	sm2_generic_format(xfer_entry, ctx->ep->gid, ctx->op, 0, msg->data,
+			   ctx->op_flags, msg->context);
+	cmd_rma = (struct sm2_cmd_sar_rma_msg *) xfer_entry->user_data;
+	memset(cmd_rma, 0, sizeof(*cmd_rma));
+	cmd_rma->sar_hdr.context = ctx;
+
+	payload_used = 0;
+	if (ctx->op == ofi_op_read_req) {
+		cmd_rma->sar_hdr.request_offset = ctx->bytes_requested;
+		ctx->bytes_requested += SM2_RMA_INJECT_SIZE;
+		if (ctx->bytes_requested >= ctx->bytes_total) {
+			ctx->bytes_requested = ctx->bytes_total;
+			cmd_rma->sar_hdr.proto_flags |=
+				FI_SM2_SAR_LAST_MESSAGE_FLAG;
+		}
+		memcpy(cmd_rma->rma_iov, msg->rma_iov,
+		       sizeof(cmd_rma->rma_iov));
+		ctx->msgs_in_flight++;
+
+		return FI_SUCCESS;
+	}
+
+	for (jv = 0; jv < SM2_IOV_LIMIT; jv++) {
+		if (ctx->bytes_sent == ctx->bytes_total)
+			break;
+		bytes_to_send =
+			MIN(msg->msg_iov[0].iov_len, msg->rma_iov[0].len);
+		bytes_to_send =
+			MIN(bytes_to_send, SM2_RMA_INJECT_SIZE - payload_used);
+		if (bytes_to_send == 0)
+			break;
+
+		src = (uint8_t *) msg->msg_iov[0].iov_base;
+		sm2_get_iface_device((void **) msg->desc, &iface, &device);
+		cmd_rma->rma_iov[jv].addr = msg->rma_iov[0].addr;
```

Comment:
if we are always going to hard code rma_iov[0] to 0, should it even be an array?

The answer is that I think we need to send the full rma_iov so the receiving side knows where to put the data.

Reply:
hinted above.  rma_iov[0] is always the right address for the remote to write into for this particular packet because we keep consuming the iov list which might fill the whole array.  Because we fill this rma list each time, the remote needs no context.

The alternative I think is to send the original, full RMA iov, then the remote needs to know which offset this packet corresponds to, does a "consume" of that many bytes, and uses the remaining values in the iov array to perform the copy of however many bytes are in the packet.  This also requires a packet size I suppose, since the RMA iov currently provides that for me.